src/profile-logic/track-initialization.js

```diff
@@ -0,0 +1,332 @@
+          pid: pid,
+          mainThreadIndex: null,
+        };
+        globalTracks.push(globalTrack);
+        globalTracksByPid.set(pid, globalTrack);
+      }
+    }
+  }
+  return globalTracks;
+}
+
+export function initializeGlobalTrackOrder(
+  globalTracks: GlobalTrack[],
+  urlGlobalTrackOrder: TrackIndex[] | null
+): TrackIndex[] {
+  return urlGlobalTrackOrder !== null &&
+    _indexesAreValid(globalTracks.length, urlGlobalTrackOrder)
+    ? urlGlobalTrackOrder
+    : globalTracks.map((_, index) => index);
+}
+
+export function initializeSelectedThreadIndex(
+  selectedThreadIndex: ThreadIndex | null,
+  visibleThreadIndexes: ThreadIndex[],
+  profile: Profile
+): ThreadIndex | null {
+  if (
+    selectedThreadIndex !== null &&
+    visibleThreadIndexes.includes(selectedThreadIndex)
+  ) {
+    // This is a valid thread index to select.
+    return selectedThreadIndex;
+  }
+  // Select either the GeckoMain [tab] thread, or the first thread in the thread
+  // order.
+  const threadIndex = profile.threads.indexOf(
+    _findDefaultThread(
+      visibleThreadIndexes.map(threadIndex => profile.threads[threadIndex])
+    )
+  );
+  // The threadIndex being null should only happen in tests where profiles
+  // have no threads.
+  return threadIndex === -1 ? null : threadIndex;
+}
+
+export function initializeHiddenGlobalTracks(
+  globalTracks: GlobalTrack[],
+  threads: Thread[],
+  validTrackIndexes: TrackIndex[],
+  urlHiddenGlobalTracks: Set<TrackIndex> | null
+): Set<TrackIndex> {
+  const hiddenGlobalTracks = new Set();
+
+  if (urlHiddenGlobalTracks === null) {
+    // No hidden global tracks exist, generate the default Set.
+    for (const track of globalTracks) {
+      if (track.type === 'process' && track.mainThreadIndex !== null) {
+        const { mainThreadIndex } = track;
+        const thread = threads[mainThreadIndex];
+        if (_isThreadIdle(thread)) {
+          hiddenGlobalTracks.add(mainThreadIndex);
+        }
+      }
+    }
+    return hiddenGlobalTracks;
+  }
+
+  // Validate the global tracks to hide.
+  for (const trackIndex of urlHiddenGlobalTracks) {
+    if (validTrackIndexes.includes(trackIndex)) {
+      hiddenGlobalTracks.add(trackIndex);
+    }
+  }
+  return hiddenGlobalTracks;
+}
+
+export function getVisibleThreads(
+  globalTracks: GlobalTrack[],
+  hiddenGlobalTracks: Set<TrackIndex>,
+  localTracksByPid: Map<Pid, LocalTrack[]>,
+  hiddenTracksByPid: Map<Pid, Set<TrackIndex>>
+): ThreadIndex[] {
+  const visibleThreads = [];
+  for (const globalTrack of globalTracks) {
+    if (globalTrack.type === 'process') {
+      const { mainThreadIndex, pid } = globalTrack;
+      if (
+        mainThreadIndex !== null &&
+        !hiddenGlobalTracks.has(mainThreadIndex)
+      ) {
+        visibleThreads.push(mainThreadIndex);
+      }
+      const tracks = localTracksByPid.get(pid);
+      const hiddenTracks = hiddenTracksByPid.get(pid);
+      if (tracks !== undefined && hiddenTracks !== undefined) {
```

Comment:
can you add a comment that this condition is for flow only, and that we're guaranteed that both aren't undefined?

I'm asking because at first glance if `hiddenTracks` is undefined but not `tracks` we could have an issue. And actuallt maybe we could have an "else" for this condition with an error throw to ensure this ?

Or actually even better: use `utils/flow`'s `ensureExists` on the 2 previous lines ?